src/basics.ml

```diff
@@ -240,6 +240,21 @@ let distinct_vars : term array -> bool =
     | _ -> raise Not_unique_var
   in try Array.iter check ts; true with Not_unique_var -> false
 
+(** [is_pattern t] tells whether term [t] is a valid pattern. *)
+let rec is_pattern : term -> bool = fun t ->
```

Comment:
I think what we want is to reject rules in which a pattern variable is applied to something.
To do that, you just need to use `Basics.get_args`, and check that the argument list is empty if the head is a pattern variable. And then call the checking function recursively on the arguments if that is not the case. I don't remember if that's enough though. You might want to lookup the exact definition of Miller patterns.

Reply:
We could allow something like `&F x y` if `x` and `y` are two bound variables. But we can enforce writing `F[x,y]`.

Reply:
Ah but it breaks compatibility with legacy syntax.

Reply:
The idea was always to compile applications of symbol `F` to distinct bound variables `x` and `y` (in the legacy syntax) as `&F[x,y]`, not as `&F x y`.

Reply:
If that is not what is done, that's a bug.

Reply:
OK, I'll have a look.

Reply:
It looks like there is a bug indeed, if I print the obtained lhs in `scope.ml`, I obtain, on `55.dk`
```
h &aa (λx, &aa x)
h (λx, &aa x) &aa
```

Reply:
No, that's normal. In the dk file, aa is applied once to no argument, and another time to one argument. In this case, we replace aa by a pattern variable &aa of the smallest arity (0 here). Changing this would require to eta-expand some patterns.

Reply:
Eta-expanding some patterns would seem more reasonable to me, since that would follow the Lambdapi approach more closely. Are we even sure that matching behaves correctly on such patterns?

Reply:
Let's keep it as it is now. There is no need to change that at the moment.